### ISSUE TITLE
Better ways of dealing with snapshots in loops or helper functions

### ISSUE THREAD

**Issue by ilyagr** [mitsuhiko/insta#313]
I have a few thoughts about https://github.com/mitsuhiko/insta/commit/745b45b5b47e5c2b60ff06ab9bf4af2138dc6d5a.

Firstly, the error message saying `Insta does not allow inline snapshot assertions in loops` isn't quite right. The new panic can happen without any loops:

```rust
fn helper(i:i64) {
   assert_snapshot!(i.to_string(), @"1");
}

#[test]
fn a_test() {
    helper(1);
    helper(2);
}
```

Here's [how I encountered this situation](https://github.com/martinvonz/jj/blob/e9952bcf76d1a25eccd414f7980d369299d02a89/tests/test_resolve_command.rs#L234-L251).

Secondly, a friendlier behavior would be to not have an error merely because the snapshot at the same line was encountered twice. Rather, `insta` could error out only if , when this snapshot is repeatedly encountered, `insta` sees a different string compared to it than it saw before. Then, tests that were correct before wouldn't start failing because of upgrading `insta`. 

In other words, the example above would have `insta` abort and report an error. But if we replace `helper` with

```rust
fn helper(i:i64) {
   do_something_complicated(i);
   assert_snapshot!(mission_report(), @"Total success!");
}
```

the test should be valid. Ideally (but not if it's too much trouble to implement), if `mission_report()` is modified to report "Complete success" both times, `insta` could handle that kind of change and allow the user to use the normal `cargo insta review` functionality to fix the test.

It's hard for me to tell how much effort it would take to make this happen. Is this the kind of thing you would accept a PR for? (No promises at this point, though) Do you think it's not worth the effort?

**Comment by mitsuhiko** [comment]
It could be changed so that it only complains if the assertion fails, but I'm not convinced that is a good idea. A snapshot test that should be consistent across multiple runs smells. What if the runs are not consistent and you get three different results? Particularly with `cargo insta test` this would create three inline update records which was not really intended.

**Comment by ilyagr** [comment]
> A snapshot test that should be consistent across multiple runs smells. 

I guess I don't think so. This is the only kind of snapshot that `insta` *can* support across multiple runs, and that seems OK to me.

>  What if the runs are not consistent and you get three different results?

The idea is that once you get two different results, you treat that the same as any test failure that insta can't fix by editing a golden test snapshot. Ideally, `insta` would report a detailed error showing both and either panic or otherwise abort the test. It shouldn't try to create any update records at this point.

**Comment by mitsuhiko** [comment]
For instance to detect all potential runs of a snapshot test it would require functionality that rust-test just does not provide. I'm not sure what it can realistically do in that scenario.

**Comment by ilyagr** [comment]
I'm not sure I understand your concern, forgive me if I'm missing some obvious flaw in my design. I wasn't suggesting that `insta` tries to predict whether a snapshot can or cannot be encountered multiple times, and I don't think that's necessary for it to work.

By the way, I'm not trying to present this as a particularly serious or urgent problem. In the one case this was a problem for me, it was just a small annoyance. The workaround of replacing `insta::assert_snapshot` with `assert_eq!` was fine. Its main downside is that it may potentially make the life of someone breaking the test slightly harder, but the difference would be slight. I have no idea if this is/can become a problem without an easy workaround for someone (or a future me).

Also, thanks for making `insta`! It's a neat and really useful tool.

**Comment by ilyagr** [comment]
You are correct about my proposal. I didn't quite understand your explanation (I didn't experience `insta` having any trouble restricting itself to snapshots from latest invocation), but I am sure I can figure it out with some research.  In any case, I accept that there is indeed some issue I haven't considered. :)

> In a way there is an easier solution I believe which is to have the helper add to a HashSet and assert on the contents of it.

That's a good point. I don't actually know how to use snapshots that are HashSet-s or vectors, as opposed to strings, but I'll look at the docs. If that's easy enough, I like this approach.

Thanks again!

**Comment by irevoire** [comment]
I was wondering about the exact same thing while rewriting our test suite with more insta everywhere.
I just wanted to share my use case without diving into the implementation details.

At meilisearch, we use insta _a lot_ to ensure we return the expected error message.
And one route we work with a lot is our `search` route. Since this route can be accessed both with GET query parameter and POST payload, we made a function that takes a `serde_json::Value` and serializes it as a query parameter and a json payload.

Since both routes are supposed to work similarly, we apply the same asserts.
Here is what it looks like in practice;
```rust
#[actix_rt::test]
async fn search_unexisting_index() {
    let server = Server::new().await;
    let index = server.index("test");

    let expected_response = json!({
        "message": "Index `test` not found.",
        "code": "index_not_found",
        "type": "invalid_request",
        "link": "https://docs.meilisearch.com/errors#index_not_found"
    });

    index
        .search(json!({"q": "hello"}), |response, code| {
            assert_eq!(code, 404);
            assert_eq!(response, expected_response);
        })
        .await;
}
```

And what I would like to write instead;
```rust
#[actix_rt::test]
async fn search_unexisting_index() {
    let server = Server::new().await;
    let index = server.index("test");

    index
        .search(json!({"q": "hello"}), |response, code| {
            assert_display_snapshot!(code, @"404");
            assert_json_snapshot!(response, @"...");
        })
        .await;
}
```

**Comment by mitsuhiko** [comment]
If that functionality were to be added, would this be an acceptable way to express that desire?

```rust
#[actix_rt::test]
async fn search_unexisting_index() {
    let server = Server::new().await;
    let index = server.index("test");

    insta::multi_snapshots! {
        index
            .search(json!({"q": "hello"}), |response, code| {
                assert_display_snapshot!(code, @"404");
                assert_json_snapshot!(response, @"...");
            })
            .await;
    }
}
```

**Comment by irevoire** [comment]
Yep definitely, both would work, I just wanted to share another "real" use case we encountered while using insta 😁

**Comment by mtbc** [comment]
For what it's worth, I just ran into the same issue. I have two tests that I want to be clear have the same first part of setup but a different conclusion. So, I had them both call the same helper to start out with, to show the difference is only toward the end. However, that helper setup uses `insta::assert...` to make sure the setup's going as expected, leaving me with that same "Insta does not allow inline snapshot assertions in loops" where there isn't a loop.

**Comment by mitsuhiko** [comment]
I want to implement this but I cannot come up with a good name for the macro that explains what it's actually doing. Here some options:

## Option A

```rust
insta::assert_perfect_duplicates! {
    for x in (0..10).step_by(2) {
        let is_even = x % 2 == 0;
        insta::assert_debug_snapshot!(is_even, @"true");
    }
}
```

## Option B

```rust
insta::assert_multi_snapshots! {
    for x in (0..10).step_by(2) {
        let is_even = x % 2 == 0;
        insta::assert_debug_snapshot!(is_even, @"true");
    }
}
```

## Option C

```rust
insta::multi_snapshots! {
    for x in (0..10).step_by(2) {
        let is_even = x % 2 == 0;
        insta::assert_debug_snapshot!(is_even, @"true");
    }
}
```

## Option D

```rust
insta::allow_matching_duplicates! {
    for x in (0..10).step_by(2) {
        let is_even = x % 2 == 0;
        insta::assert_debug_snapshot!(is_even, @"true");
    }
}
```

## Option E

```rust
insta::allow_duplicates! {
    for x in (0..10).step_by(2) {
        let is_even = x % 2 == 0;
        insta::assert_debug_snapshot!(is_even, @"true");
    }
}
```

**Comment by mitsuhiko** [comment]
I'm a bit conflicted on the naming still. I at the moment think prefer `allow_duplicates` because it states that in that block duplicates are permitted. But not sure.

**Comment by max-sixty** [comment]
Naive question — would using the existing `insta::with_settings!` approach not work here? Possibly it's _only_ for scoped sections and so making it a normal setting would be confusing?

**Comment by mitsuhiko** [comment]
It maintains a stack of previous run results. I would not want to abuse the settings for that since they are not necessarily scope bound.

**Comment by ilyagr** [comment]
`allow_reentrant_snapshots` is a bit verbose but descriptive.

**Comment by mitsuhiko** [comment]
@ilyagr reentrancy means that a function can be invoked again before it finished. This is not what is happening here so I think the name would not be correct.

**Comment by ilyagr** [comment]
`allow_snapshot_reexecution`?
